test/distributed/pipelining/test_schedule.py

```diff
@@ -427,7 +560,297 @@ def test_send_recv(self, test_info):
+
+        # Uncomment to regenerate reference output
+        # _dump_csv(comms_sch, comms_csv)
+
+        sch_ref = {}
+        with open(comms_csv, newline="") as ref:
+            for rank, row in enumerate(csv.reader(ref)):
+                sch_ref[rank] = [_Action.from_str(s) for s in row]
+
+        for rank in sch_ref:
+            for timestep, (a, b) in enumerate(zip(comms_sch[rank], sch_ref[rank])):
+                self.assertEqual(a, b, f"Mismatch at {timestep=}, {a=}, expected {b}")
+
+        simulated_schedule = _simulate_comms_compute(
+            comms_sch,
+            stage_to_rank=lambda s: s % pipeline_parallel_size,
+            num_stages=num_stages,
+        )
+
+        num_steps = max([len(simulated_schedule[rank]) for rank in simulated_schedule])
+        # print(_format_pipeline_order(simulated_schedule))
+        self.assertEqual(num_steps, 113)
+
+    @requires_nccl()
+    def test_grad_with_v_schedule(self):
+        """
+        We have a special case for V schedules where 2 adjacent stages are on the same rank.
+        E.g.
+        rank0:  stage 0,                 stage3
+        rank1:          stage 1, stage 2,
+
+        The special case involves not using send/recv ops but directly passing tensors between colocated stages.
+
+        This test runs on a single rank and just tests the 'stage1, stage2' portion for both F and B, comparing
+        gradients to a reference model with 2 layers.
+        """
+        store = FakeStore()
+        torch.distributed.init_process_group(
+            backend="fake", rank=0, world_size=1, store=store
+        )
+        d_hid = 512
+        batch_size = 256
+        n_stages = 2
+        device = "cuda"
+        full_mod = MultiMLP(d_hid, n_layers=n_stages)
+        full_mod.to(device)
+
+        ref_mod = copy.deepcopy(full_mod)
+        x = torch.randn(batch_size, d_hid, device=device)
+        with torch.no_grad():
+            y = ref_mod(x)
+            # Add a small perturbation
+            target = y + torch.randn(batch_size, d_hid, device=device)
+
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        # Run reference
+        for _ in range(2):
+            ref_mod.zero_grad()
+            ref_out = ref_mod(x)
+            ref_loss = loss_fn(ref_out, target)
+            ref_loss.backward()
+
+        stage_indices = [0, 1]
+        submod_names = [f"layers.{i}" for i in stage_indices]
+        stage_modules = [
+            full_mod.get_submodule(submod_name) for submod_name in submod_names
+        ]
+        # Create a pipeline stage to wrap that submodule
+        num_microbatches = 2
+        input_args = x.chunk(num_microbatches)[0]
+        stages = [
+            PipelineStage(
+                stage_module,
+                stage_idx,
+                n_stages,
+                device,
+                # TODO(whc) shape inference shouldn't have needed to run communications in this 1-rank, 2-stage scenario,
```

Comment:
removed this, it is fixed now. probably fixed by @H-Huang's PR to fix single-stage schedule usage.